### PR TITLE
Add support for function annotation

### PR DESCRIPTION
**Pull request by GeorgeSittas**

### CODE REVIEW

File: sqlglot/optimizer/annotate_types.py [hunk 1]
```diff
@@ -104,6 +108,78 @@ class TypeAnnotator:
         },
     }
 
+    FUNCTION_TYPES = {
```
Comment (barakalon) [inline sqlglot/optimizer/annotate_types.py:111]: What about functions like `COERCE`, in which the return type depends on the input type?

Comment (tobymao) [reply sqlglot/optimizer/annotate_types.py:111]: it should be an optional lambda

Comment (barakalon) [reply sqlglot/optimizer/annotate_types.py:111]: Ah, we're expecting "custom annotator functions" for this.

Comment (tobymao) [reply sqlglot/optimizer/annotate_types.py:111]: why do you need FUNCTION_TYPES dict at all? why not just use annotators?

Comment (GeorgeSittas) [reply sqlglot/optimizer/annotate_types.py:111]: That's what I'm working on, pushing in a bit.

**Comment by tobymao** [comment]
anything with null is in general null

**Comment by GeorgeSittas** [comment]
Outstanding items for this PR:

- [ ] Find a way to handle Nullable expressions (eg. columns, and/or expressions). Should probably use the `Nullable` type.

File: sqlglot/optimizer/annotate_types.py [hunk 2]
```diff
@@ -141,36 +211,45 @@ def _maybe_annotate(self, expression):
...
 
     def _annotate_binary(self, expression):
         self._annotate_args(expression)
 
-        if isinstance(expression, (exp.Condition, exp.Predicate)):
+        left_type = expression.left.type
+        right_type = expression.right.type
+
+        if isinstance(expression, (exp.And, exp.Or)):
+            if (left_type, right_type) == (exp.DataType.Type.NULL, exp.DataType.Type.NULL):
```
Comment (tobymao) [inline sqlglot/optimizer/annotate_types.py:243]: just do left_type == null and

File: sqlglot/optimizer/annotate_types.py [hunk 3]
```diff
@@ -141,36 +211,45 @@ def _maybe_annotate(self, expression):
...
+                expression.type = exp.DataType.Type.NULL
+            elif exp.DataType.Type.NULL in (left_type, right_type):
+                expression.type = exp.DataType(this=exp.DataType.Type.NULLABLE, expressions=[exp.DataType.Type.BOOLEAN])
```
Comment (tobymao) [inline sqlglot/optimizer/annotate_types.py:246]: DataType.build("NULLABLE", expressions=DataType.build("BOOLEAN")

File: sqlglot/tokens.py [hunk 4]
```diff
@@ -579,6 +579,7 @@ class Tokenizer(metaclass=_Tokenizer):
         "INT8": TokenType.BIGINT,
         "DECIMAL": TokenType.DECIMAL,
         "MAP": TokenType.MAP,
+        "NULLABLE": TokenType.NULLABLE,
```
Comment (tobymao) [inline sqlglot/tokens.py:582]: you can delete the override in clickhouse with this change

**Pull request merged**